### PR TITLE
ImportApi operation for apigatewayv2

### PR DESCRIPTION
**Pull request by vijtrip2**
Issue #128 , if available:

Description of changes:
* **Fixing the classic e2e test run. Temporarily commenting docker test run.**
    * Update ack-generate/release to drop ".tpl" from files generated for helm charts.
    * I spent one day trying to fix docker-file and scripts for successfully running docker-test-run but in the end I hit the blocker where daws was unable to run inside docker.
    * Then I fall back to old e2e runs and temporarily commented docker-test-runs from kind-build-test
    * I made some minor fixes in script for successfully running e2e tests. 

* **Adding required override for fields in FieldConfig.**
    * Introduced a new boolean field 'required' in FieldConfig to override the required behavior present in aws-sdk-go.
    * This was required for ImportApi operation to succeed without taking required-fields from CreateApi operation.

* **Implementing ImportApi & ReimportApi for apigatewayv2.**
    * Created custom methods for Import and Reimport Api operations.
    * Updated e2e tests for apigatewayv2 to include ImportApi and ReimportApi.

e2e test run was successful.
```
vijat@186590cff3cf aws-controllers-k8s % ./scripts/kind-build-test.sh apigatewayv2
checking AWS credentials ... ok.
creating kind cluster ack-test-2b2ebd8b-d02fcaf7 ... ok.
[+] Building 93.8s (17/17) FINISHED                                                                                                          
 => [internal] load build definition from Dockerfile                                                                                    0.0s
 => => transferring dockerfile: 39B                                                                                                     0.0s
 => [internal] load .dockerignore                                                                                                       0.0s
 => => transferring context: 2B                                                                                                         0.0s
 => [internal] load metadata for docker.io/library/amazonlinux:2                                                                        1.5s
 => [internal] load metadata for docker.io/library/golang:1.14.1                                                                        1.5s
 => [auth] library/amazonlinux:pull token for registry-1.docker.io                                                                      0.0s
 => [auth] library/golang:pull token for registry-1.docker.io                                                                           0.0s
 => [internal] load build context                                                                                                       1.5s
 => => transferring context: 809.40kB                                                                                                   1.5s
 => [builder 1/7] FROM docker.io/library/golang:1.14.1@sha256:38a520b05b971181a5a30d72b74d1b70c63688eb267a4fb376c7476a5c6f885b          0.0s
 => => resolve docker.io/library/golang:1.14.1@sha256:38a520b05b971181a5a30d72b74d1b70c63688eb267a4fb376c7476a5c6f885b                  0.0s
 => CACHED [stage-1 1/3] FROM docker.io/library/amazonlinux:2@sha256:e87ffde799fd6e9c70c6ffb190726c146aeb1ae4d7667848850d9a609e7a7ab0   0.0s
 => CACHED [builder 2/7] WORKDIR /github.com/aws/aws-controllers-k8s                                                                    0.0s
 => CACHED [builder 3/7] COPY go.mod go.mod                                                                                             0.0s
 => CACHED [builder 4/7] COPY go.sum go.sum                                                                                             0.0s
 => CACHED [builder 5/7] RUN  go mod download                                                                                           0.0s
 => [builder 6/7] COPY . /github.com/aws/aws-controllers-k8s/                                                                           8.9s
 => [builder 7/7] RUN GIT_VERSION=$(git describe --tags --dirty --always) &&     GIT_COMMIT=$(git rev-parse HEAD) &&     BUILD_DATE=$  79.9s
 => [stage-1 2/3] COPY --from=builder /github.com/aws/aws-controllers-k8s/bin/controller /github.com/aws/aws-controllers-k8s/LICENSE /  0.4s
 => exporting to image                                                                                                                  0.4s
 => => exporting layers                                                                                                                 0.4s
 => => writing image sha256:841c676bcbe3a3ddad6393b7bc65d8b8dc47dcd1fd81a188e44526b892386f33                                            0.0s
 => => naming to docker.io/library/aws-controllers-k8s:apigatewayv2-v0.0.1-119-g735f9da-dirty                                           0.0s
ok.
loading the images into the cluster ... ok.
loading CRD manifests for apigatewayv2 into the cluster ... ok.
loading RBAC manifests for apigatewayv2 into the cluster ... ok.
loading service controller Deployment for apigatewayv2 into the cluster ...ok.
generating AWS temporary credentials and adding to env vars map ... ok.
======================================================================================================
To poke around your test cluster manually:
export KUBECONFIG=/Users/vijat/Documents/gocode/src/github.com/vijtrip2/aws-controllers-k8s/scripts/lib/../../build/tmp-ack-test-2b2ebd8b-d02fcaf7/kubeconfig
kubectl get pods -A
======================================================================================================
testing Helm release for apigatewayv2 for release version v0.0.1-119-g735f9da-dirty.
Building release artifacts for apigatewayv2-v0.0.1-119-g735f9da-dirty
Generating custom resource definitions for apigatewayv2
Generating RBAC manifests for apigatewayv2
namespace/ack-system-test-helm created
installing the helm chart for ack-apigatewayv2-controller in namespace ack-system-test-helm ... ok.
uninstalling the helm chart for ack-apigatewayv2-controller in namespace ack-system-test-helm ... ok.
e2e took 192 second(s)
🥑 Deleting k8s cluster using "kind"
Deleting cluster "ack-test-2b2ebd8b-d02fcaf7" ...
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by jaypipes — CHANGES_REQUESTED** [review]
Few small requests inline, please see my comments, @vijtrip2. Overall this looks very solid, though.

One thing you will need to do in a followup PR is create e2e tests for the failure scenarios (in particular, the Bad Input scenarios). You have created unit tests that adequately test the Go custom update and create methods themselves, but you will want to add TerminalConditions the CR's `Status.Conditions` collection and mark the CR as being unable to be synced.

See here for an example of where Elasticache is doing this:

https://github.com/aws/aws-controllers-k8s/blob/main/services/elasticache/generator.yaml#L7-L11

which generates this code:

https://github.com/aws/aws-controllers-k8s/blob/main/services/elasticache/pkg/resource/cache_subnet_group/sdk.go#L387-L399

And here is an example of the Elasticache Snapshot resource throwing a terminal condition in its custom create code paths:

https://github.com/aws/aws-controllers-k8s/blob/main/services/elasticache/pkg/resource/snapshot/custom_create_api.go#L31-L32

and that exception code is listed in the generator here:

https://github.com/aws/aws-controllers-k8s/blob/main/services/elasticache/generator.yaml#L59

Ping @nmvk for details on how the TerminalCondition stuff works :)

File: cmd/ack-generate/command/release.go [hunk 1]
```diff
@@ -108,6 +108,9 @@ func generateRelease(cmd *cobra.Command, args []string) error {
 			fmt.Println(strings.TrimSpace(contents.String()))
 			continue
 		}
+		if filepath.Ext(path) == ".tpl" {
```
[the remark this answers is not in the document]
Comment (vijtrip2) [reply cmd/ack-generate/command/release.go:111]: cherry-picked the commit.

File: pkg/generate/config/field.go [hunk 2]
```diff
@@ -48,4 +48,7 @@ type FieldConfig struct {
 	// From instructs the code generator that the value of the field should
 	// be retrieved from the specified operation and member path
 	From *SourceFieldConfig `json:"from,omitempty"`
+	// Required indicates whether this field is a required member or not.
+	// This field is used to configure '+kubebuilder:validation:Required' on API object's members.
+	Required *bool `json:"required,omitempty"`
```
[the remark this answers is not in the document]
Comment (vijtrip2) [reply pkg/generate/config/field.go:53]: Done.

File: pkg/model/crd.go [hunk 3]
```diff
@@ -248,7 +254,8 @@ func (r *CRD) AddSpecField(
 	memberNames names.Names,
 	shapeRef *awssdkmodel.ShapeRef,
 ) *CRDField {
-	crdField := newCRDField(r, memberNames, shapeRef, nil)
+  fieldConfigs := r.genCfg.ResourceFields(r.Names.Original)
```
[the remark this answers is not in the document]
Comment (vijtrip2) [reply pkg/model/crd.go:257]: Done. I fixed conflict on github-ui. That's why this become unindented. Fixed now.

File: scripts/lib/aws/apigatewayv2.sh [hunk 4]
```diff
@@ -46,6 +46,114 @@ EOF
...
+# import_http_api_and_validate accepts one required parameter api_name
+apigwv2_import_http_api_and_validate() {
+  if [[ $# -ne 1 ]]; then
+    echo "FATAL: Wrong number of arguments passed to apigwv2_import_http_api_and_validate"
+    echo "Usage: apigwv2_import_http_api_and_validate api_name"
+    exit 1
+  fi
+
+  local __api_name="$1"
+  local __api_resource_name=api/"$1"
```
[the remark this answers is not in the document]
Comment (vijtrip2) [reply scripts/lib/aws/apigatewayv2.sh:60]: Nope. api/"$1" works fine. Just similar to apigwv2_create_http_api_and_validate

**Comment by vijtrip2** [comment]
> Few small requests inline, please see my comments, @vijtrip2. Overall this looks very solid, though.
> 
> One thing you will need to do in a followup PR is create e2e tests for the failure scenarios (in particular, the Bad Input scenarios). You have created unit tests that adequately test the Go custom update and create methods themselves, but you will want to add TerminalConditions the CR's `Status.Conditions` collection and mark the CR as being unable to be synced.
> 
> See here for an example of where Elasticache is doing this:
> 
> https://github.com/aws/aws-controllers-k8s/blob/main/services/elasticache/generator.yaml#L7-L11
> 
> which generates this code:
> 
> https://github.com/aws/aws-controllers-k8s/blob/main/services/elasticache/pkg/resource/cache_subnet_group/sdk.go#L387-L399
> 
> And here is an example of the Elasticache Snapshot resource throwing a terminal condition in its custom create code paths:
> 
> https://github.com/aws/aws-controllers-k8s/blob/main/services/elasticache/pkg/resource/snapshot/custom_create_api.go#L31-L32
> 
> and that exception code is listed in the generator here:
> 
> https://github.com/aws/aws-controllers-k8s/blob/main/services/elasticache/generator.yaml#L59
> 
> Ping @nmvk for details on how the TerminalCondition stuff works :)

Thank you very much. I will incorporate this into upcoming PR. Pasting these very good instructions into issue #128

**Comment by vijtrip2** [comment]
Recent successful e2e run.

```
vijat@186590cff3cf aws-controllers-k8s % ./scripts/kind-build-test.sh apigatewayv2
checking AWS credentials ... ok.
creating kind cluster ack-test-06a6e1e9-89c2341e ... ok.
[+] Building 159.9s (17/17) FINISHED                                                                                                         
 => [internal] load build definition from Dockerfile                                                                                    0.1s
 => => transferring dockerfile: 39B                                                                                                     0.0s
 => [internal] load .dockerignore                                                                                                       0.0s
 => => transferring context: 2B                                                                                                         0.0s
 => [internal] load metadata for docker.io/library/amazonlinux:2                                                                        2.4s
 => [internal] load metadata for docker.io/library/golang:1.14.1                                                                        1.6s
 => [auth] library/golang:pull token for registry-1.docker.io                                                                           0.0s
 => [auth] library/amazonlinux:pull token for registry-1.docker.io                                                                      0.0s
 => [builder 1/7] FROM docker.io/library/golang:1.14.1@sha256:38a520b05b971181a5a30d72b74d1b70c63688eb267a4fb376c7476a5c6f885b          0.0s
 => => resolve docker.io/library/golang:1.14.1@sha256:38a520b05b971181a5a30d72b74d1b70c63688eb267a4fb376c7476a5c6f885b                  0.0s
 => [stage-1 1/3] FROM docker.io/library/amazonlinux:2@sha256:420a6503505c73ea05e1254acd4d1a5e73305dab8ec3fa70833e112cc6980624         20.4s
 => => resolve docker.io/library/amazonlinux:2@sha256:420a6503505c73ea05e1254acd4d1a5e73305dab8ec3fa70833e112cc6980624                  0.0s
 => => sha256:420a6503505c73ea05e1254acd4d1a5e73305dab8ec3fa70833e112cc6980624 547B / 547B                                              0.0s
 => => sha256:1c5213077930ebbd35b51bece063ffb31f47df9c6fb71b7248944785c77ba772 529B / 529B                                              0.0s
 => => sha256:9917dfe117ca1e8affbea6d819c0d94918524ae1882fe69746a8217aec4c4053 1.48kB / 1.48kB                                          0.0s
 => => sha256:4232b9ee675bb1f9a32337ec3f266a570b75c54c3c3f173b1abb998e9984ad8c 62.01MB / 62.01MB                                        5.7s
 => => extracting sha256:4232b9ee675bb1f9a32337ec3f266a570b75c54c3c3f173b1abb998e9984ad8c                                              13.8s
 => [internal] load build context                                                                                                       7.7s
 => => transferring context: 20.10MB                                                                                                    7.5s
 => CACHED [builder 2/7] WORKDIR /github.com/aws/aws-controllers-k8s                                                                    0.0s
 => CACHED [builder 3/7] COPY go.mod go.mod                                                                                             0.0s
 => CACHED [builder 4/7] COPY go.sum go.sum                                                                                             0.0s
 => CACHED [builder 5/7] RUN  go mod download                                                                                           0.0s
 => [builder 6/7] COPY . /github.com/aws/aws-controllers-k8s/                                                                          17.9s
 => [builder 7/7] RUN GIT_VERSION=$(git describe --tags --dirty --always) &&     GIT_COMMIT=$(git rev-parse HEAD) &&     BUILD_DATE=  127.6s
 => [stage-1 2/3] COPY --from=builder /github.com/aws/aws-controllers-k8s/bin/controller /github.com/aws/aws-controllers-k8s/LICENSE /  1.1s
 => exporting to image                                                                                                                  1.4s
 => => exporting layers                                                                                                                 1.4s
 => => writing image sha256:b1fd2407d2d1abbf74d16f932c165a0bcb42665101c054203249d1620a06e1d9                                            0.0s
 => => naming to docker.io/library/aws-controllers-k8s:apigatewayv2-v0.0.1-127-gc15e700-dirty                                           0.0s
ok.
loading the images into the cluster ... ok.
loading CRD manifests for apigatewayv2 into the cluster ... ok.
loading RBAC manifests for apigatewayv2 into the cluster ... ok.
loading service controller Deployment for apigatewayv2 into the cluster ...ok.
generating AWS temporary credentials and adding to env vars map ... ok.
======================================================================================================
To poke around your test cluster manually:
export KUBECONFIG=/Users/vijat/Documents/gocode/src/github.com/vijtrip2/aws-controllers-k8s/scripts/lib/../../build/tmp-ack-test-06a6e1e9-89c2341e/kubeconfig
kubectl get pods -A
======================================================================================================
testing Helm release for apigatewayv2 for release version v0.0.1-127-gc15e700-dirty.
Building release artifacts for apigatewayv2-v0.0.1-127-gc15e700-dirty
Generating custom resource definitions for apigatewayv2
Generating RBAC manifests for apigatewayv2
namespace/ack-system-test-helm created
installing the helm chart for ack-apigatewayv2-controller in namespace ack-system-test-helm ... ok.
uninstalling the helm chart for ack-apigatewayv2-controller in namespace ack-system-test-helm ... ok.
e2e took 220 second(s)
🥑 Deleting k8s cluster using "kind"
Deleting cluster "ack-test-06a6e1e9-89c2341e" ...
```

**Review by jaypipes — COMMENTED** [review]
@vijtrip2 please rebase and squash your commits into one please. Otherwise, this is good to go!

**Comment by vijtrip2** [comment]
closing in favor of PR #609

**Pull request closed without merging**